src/message_dao.py

```diff
@@ -109,6 +108,26 @@ def get(self, message_id: str) -> Optional[Message]:
         msg.children = self._get_children(message_id)
         return msg
 
+    def find_matching(self, normalized_subject: str = "", from_: str = "") -> List[Message]:
+        criteria = {
+            "normalized_subject": normalized_subject,
+            "from_": from_
+        }
+    
+        non_empty = [(attr, value) for attr, value in criteria.items() if value != '']
+        if len(non_empty) == 0:
+            raise RuntimeError("Need to specify some attribute to match on!")
+        
+        clauses = [f' {attr}=%s' for attr, _ in non_empty]
+        values = [value for _, value in non_empty]
+
+        query = "SELECT message_id FROM Messages WHERE change_id IS NOT NULL AND" + " AND".join(clauses)
+
+        with self.connection.cursor() as cursor:
+            cursor.execute(query, tuple(values))
+            res = cursor.fetchall()
+        return [self.get(message_id) for message_id in (tup[0] for tup in res)]
```

Comment:
we have a double comprehension going on.

Would it be simpler as 
```python
[self.get(tup[0]) for tup in res]
```

Alternatively, we could do
```python
[self.get(msg_id) for (msg_id,) in res]
```

but I think explicitly doing `tup[0]` is clearer than unpacking a single-element tuple

Reply:
That'd certainly be simpler. I was trying to do it the 2nd way for clearer method calls, but I didn't realize we could unpack directly like that. I'll stick with the first way that you suggested.